File: sdk/symbol_dart_sdk/bin/nem/SharedKey.dart
import 'dart:typed_data';
import '../CryptoTypes.dart';
import '../symbol/KeyPair.dart';
import '../crypto/tweetNacl.dart' as tweet_nacl;
import '../SharedKey.dart';
import 'package:pointycastle/export.dart' as pointy_castle;

void cryptoHash(Uint8List out, Uint8List m) {
  var digest = pointy_castle.KeccakDigest(512);
  var r = digest.process(m);

  for (var i = 0; i < out.length; ++i) {
    out[i] = r[i];
  }
}

final deriveSharedKeyImpl = deriveSharedKeyFactory('catapult', tweet_nacl.TweetNaCl.crypto_hash);

Future<SharedKey256> deriveSharedKey(dynamic keyPair, PublicKey otherPublicKey) {
  return deriveSharedKeyImpl(keyPair.privateKey.bytes.reversed, otherPublicKey);
}

File: sdk/csharp/SymbolSdk/Symbol/SharedKey.cs
namespace SymbolSdk.Symbol;

public abstract class SharedKey
{
    public static SharedKey256 DeriveSharedKey(byte[] privateKey, byte[] otherPublicKey) {
        var deriveSharedKeyImpl = SymbolSdk.SharedKey.DeriveSharedKeyFactory("catapult", NaclCatapult.CryptoHash);
        return deriveSharedKeyImpl(privateKey, otherPublicKey);
    }
}